lua-mode.el

```diff
@@ -1754,6 +1848,151 @@ If `lua-process' is nil or dead, start a new process first."
+	   "},{"
+	   ,(string-join
+	     (mapcar (lambda (l) (apply 'format "{var='%s',lib='%s'}" l)) libs) ",")
+	   "},{"
+	   ,(string-join (mapcar (apply-partially 'format "'%s'") locals) ",")
+	   "}," ,(number-to-string lua-shell-maximum-completions) ")")))
+
+(defvar lua-local-require-completions t
+  "During completion, scan file for local require calls for context.")
+
+(defvar lua-local-require-regexp
+  "^local\\s-+\\([^ \n]+\\)\\s-*=\\s-*require\\s-*(?\\s-*\\(['\"]?\\)\\([^)\n]+\\)\\2"
+  "A regexp to match lines where a library is required and put in a local.")
+
+(defvar lua-top-level-local-regexp
+  "^local\\s-+\\([^ \n]+\\)\\s-*="
+  "A regexp to match top-level local definitions")
+
+(defun lua-local-libs ()
+  "Find all modules loaded with require which are stored in locals.
+
+Returns a list of lists where the car is the local name and the cadr
+is the string that is passed to require."
+  (save-excursion
+    (when lua-local-require-completions
+      (let ((libs nil))
+        (goto-char (point-min))
+        ;; find each match of "local x = require y" and save for later
+        (while (lua-find-regexp 'forward lua-local-require-regexp)
+          (add-to-list 'libs (list (match-string-no-properties 1)
+                                   (match-string-no-properties 3))))
+        libs))))
+
+(defun lua-top-level-locals (lib-names)
+  "Return a list of all top-level locals in the file for completion targets."
+  (save-excursion
+    (let ((locals nil))
+      (goto-char (point-min))
+      (while (lua-find-regexp 'forward lua-top-level-local-regexp)
+        (let ((local (match-string-no-properties 1)))
+          (when (not (member local lib-names))
+            (add-to-list 'locals local))))
+      locals)))
+
+(defun lua-start-of-expr ()
+  "Search backwards to find the beginning of the current expression.
+This is distinct from `backward-sexp' which treats . and : as a separator."
+  (save-excursion
+    (backward-sexp)
+    (let ((bos (point)))
+      (when (> (point) 1) (backward-char))
+      (when (string-match "[[:space:]]" (thing-at-point 'char))
+        (search-backward-regexp "[^\n\s-]" nil t)
+        (when (string= (thing-at-point 'char) "\n")
+          (backward-char)))
+      (if (member (thing-at-point 'char) '(":" "."))
+          (lua-start-of-expr)
+        bos))))
+
+(defun lua-finalize-output ()
+  "Callback for comint-redirect-send-command
+Reads and sets output from lua-shell-output-buffer, 
+after clearing any copy of the input from beginning."
+  (setq lua-shell-redirected-output
+	(with-current-buffer lua-shell-output-buffer
+	  (buffer-substring-no-properties
+	   (point-min) (point-max)))))
+
+(defun lua-mimic-whitespace (string completions)
+  "Reproduce the whitespace pattern of the initial completion string. 
+Lua ignores whitespace, so we must complete across it.  Assumes
+string does not start with whitespace and that all completions start with
+the string."
```

Comment:
I thought that when you have `foo[bar].baz.qux.quu` and you want to complete `quu` your list of completions should have completions starting with `quu` only. At least that's what I'm used to when I use completions in other modes. Is this mimic-whitespace thing related to caching?

Reply:
This is so you can complete things like:

```lua
foo[bar].
   baz.qux
.q[M-Tab]
```

which is perfectly valid lua.  Since completion looks for exact matches (whitespace and all), we first remove the whitespace, perform the completion, then reproduce the original.

Reply:
Yes, it is valid lua. What I'm asking is a bit different.

Given `foo.b` you can either complete as "short completions", `["bar", "baz"]`, anchored at the beginning of `b` part, or you can return "long completions", `["foo.bar", "foo.baz"]`, anchored at the beginning of `foo`. Your approach clearly goes for the second way, because the first one avoids the "mimic whitespace" problem altogether, because it only returns the last component which cannot contain whitespace. "short completions" approach seems simpler, and I vaguely remember there being a reason for doing "long completions", just can't remember what was it.

Reply:
Aha, I see.  Yes, the whitespace fixing is a side effect of supporting long completions.  But you still have to identify the entire expression (whitespace and all).  Short completions were actually (silently) re-querying lua and failing before, since one of the operations of the completions framework is to ask "is this completion string "abc.def" a completion for original expression "abc"?  But the real reason is to make caching and other completions styles possible.  Completions in this context need both a fragment being completed ("abc"), and the list of completions ("abc.def", "abc.xyz").  Making the fragment an empty string means you can't cache it or use other non-trivial styles.

_(below I mistakenly thought you were referring to the depth of the search with short vs. long, but leave it here as supplemental explanation:)_
Short completions are a subset of long completions, and long completions (ideally _every possible completion below this point_, which is what all the completion docs urge you on to) are the preferred way to do completions to work with other styles.  Give the `initials` completion style a try to see what I mean.  It's pretty cool shorthand (especially with company-mode), which simply would not work with "single key at a time" shell querying.  Caching also works best when you pre-cache the `a.b.c.d.e`'s.

Reply:
Ok, let's see how it works in the wild.